.github/workflows/performance.yml

```diff
@@ -0,0 +1,160 @@
+name: Performance Tests
+
+on:
+  push:
+    branches:
+      - trunk
+      - '3.[7-9]'
+      - '[4-9].[0-9]'
+    tags:
+      - '[0-9]+.[0-9]'
+      - '[0-9]+.[0-9].[0-9]+'
+  pull_request:
+    branches:
+      - trunk
+      - '3.[7-9]'
+      - '[4-9].[0-9]'
```

Comment:
Why do we run this workflow in point release branches, when (at least for this MVP) we only deploy results for `trunk`?

Unless we log the performance results for the other branches _somewhere_, that seems like unnecessary overhead to me. Maybe I'm missing something? Otherwise I would suggest to keep only `trunk` in here.

Reply:
Even though we're only recording commits from `trunk` to the dashboard for now, I think there's value in having these measurements run all release branches, so we can inspect the action logs when suspect a performance regression has taken place in a point release. Given that commits to release branches are not as frequent, and this workflow is much faster than many others that run on each commit, I don't anticipate this adding a lot of overhead.

Long term, it would be nice to be able to visualize specific releases in dashboard as well, but that is a current limitation of the way the dashboard visualizes commits.

Reply:
Good call out.

Workflows will only run when they exist within the context of the branch being updated. So unless this workflow is backported to WP <= `6.0`, it would not run anyway.

But, I do think that this is a bit too broad. Since there are no plans to backport this branch to older versions, I think we should change this to be "from this point forward".  Similar to what we do in [newer workflows like the PHP Compatibility one](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/php-compatibility.yml#L5-L19) like this:

```
  push:
    branches:
      - trunk
      - '6.[1-9]'
      - '[7-9].[0-9]'
    tags:
      - '[6-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
  pull_request:
    branches:
      - '[0-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
      - '![0-5].[0-9].[0-9]+'
      - '!6.0.[0-9]+'
```

Reply:
@joemcgill Okay, if we are logging the results to the action logs, I agree that that is enough of a justification to have the workflow running for those branches too. 👍

Reply:
Good call out.

Workflows will only run when they exist within the context of the branch being updated. So unless this workflow is backported to WP <= `6.0`, it would not run anyway.

But, I do think that this is a bit too broad. Since there are no plans to backport this branch to older versions, I think we should change this to be "from this point forward".  Similar to what we do in [newer workflows like the PHP Compatibility one](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/php-compatibility.yml#L5-L19) like this:

```
  push:
    branches:
      - trunk
      - '6.[1-9]'
      - '[7-9].[0-9]'
    tags:
      - '[6-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
  pull_request:
    branches:
      - '[0-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
      - '![0-5].[0-9].[0-9]+'
      - '!6.0.[0-9]+'
```

Reply:
Sorry, I was mid-review and then GitHub wouldn't let me submit this one comment without submitting the whole review. And then it didn't show me the additional comments above.

> I think there's value in having these measurements run all release branches

I think this is a good point, but we should chat about tradeoffs/maintenance burden of this. Once the workflow is backported we need to backport all updates needed to keep it running.

Reply:
@desrosj We just commented opposing things at the same time 😆 

I think we may want to eventually get performance data for older branches as well; not necessarily for super old branches, but there's arguably value in making sure that e.g. 6.1.2 is not slower than 6.1.1 even once 6.2 is out. So I think this is something worth pursuing in the future.

> Workflows will only run when they exist within the context of the branch being updated. So unless this workflow is backported to WP <= `6.0`, it would not run anyway.

Is that true? @joemcgill and I had a lengthy conversation about that the other day and based on that I would think that this workflow will run for the other branches too simply because it is in the GitHub repo's main branch `trunk`.

Reply:
> I think this is a good point, but we should chat about tradeoffs/maintenance burden of this. Once the workflow is backported we need to backport all updates needed to keep it running.

That's a good point as well, though it depends on my last question above and how GitHub workflows actually apply :)

Reply:
@desrosj good point about needing to backport this to old branches. I think your suggestion to keep this to future branches makes a lot of sense. 

@felixarntz IIRC, the conversation we were having about needing a workflow in the main branch was specifically for a workflow that was meant to run on a cron schedule rather than on a branch event. Hopefully that clarifies things a bit.

Reply:
I think our answer is [here](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#about-workflow-triggers).

> 1. An event occurs on your repository. The event has an associated commit SHA and Git ref.
> 2. GitHub searches the `.github/workflows` directory in your repository for workflow files that are present in the associated commit SHA or Git ref of the event.
> 3. A workflow run is triggered for any workflows that have `on:` values that match the triggering event. Some events also require the workflow file to be present on the default branch of the repository in order to run.

In our scenario, the ref will be the branch. So if the workflow is not present in that branch, it will not run.

When we [first introduced the workflows into Core](https://core.trac.wordpress.org/browser/trunk/.github/workflows/phpunit-tests.yml?rev=49162), we basically allowed `push` events to all branches using `*.*` and to my recollection, the workflows did not run in the older branches because the workflow files were not present. It's possible this has changed, though.

Reply:
> @felixarntz IIRC, the conversation we were having about needing a workflow in the main branch was specifically for a workflow that was meant to run on a cron schedule rather than on a branch event. Hopefully that clarifies things a bit.

Yes this is accurate. `schedule` events _only_ run on the primary branch. We had to create a way around this for testing older branches receiving security updates regularly. To get around this, we have a `schedule` workflow in `trunk` that uses the GitHub API to create `workflow_dispatch` [calls to each branch for each supported workflow](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/test-old-branches.yml) in that branch.

After `6.2` is branched, it probably makes sense to add this one to this matrix as well so that it runs at least once a month.

Reply:
@joemcgill Ahhh thanks for clarifying, makes sense.

In that case, I'd indeed say we should rather remove the other branches from this PR for now.

Reply:
> After `6.2` is branched, it probably makes sense to add this one to this matrix as well so that it runs at least once a month.

I've created #4160 so we don't forget to add this.

Reply:
I think this conversation can probably be marked resolved. There were a few other inline suggestions I added to further tweak the branch and tag filtering that we can address those individually.